### DISCUSSION TITLE
Docker buildx?

### DISCUSSION THREAD

**Discussion by gotson** [jreleaser/jreleaser#1045] [Q&A] [answered]
Hi,

i read the doc (maybe not thoroughly enough) and i can't find how to build Docker image for various architectures. This is usually done with `docker buildx --platforms <platforms>`.

Is it possible with JReleaser ?

**Comment by aalmiray** [comment, marked as answer]
It's not possible yet, additional support would have to be added. Given that I have not used `buildx` before I have some additional questions:

- How would this play out with the base image? Is there a need for different base images per platform?
- Is it possible to check if `buildx` is installed/enabled? If not, can it be installed on the fly?

**Reply by gotson** [reply]
> * How would this play out with the base image? Is there a need for different base images per platform?

Usually base images are available for multiple architectures, for example https://hub.docker.com/layers/library/eclipse-temurin/17-jre/images/sha256-75141b1414c0fb4d5c6d8f2ea76fd6fc01e894f41ebf6cc53395df84023c2cba?context=explore is available on:
<img width="219" alt="image" src="https://user-images.githubusercontent.com/2139133/201604407-20143648-5750-46bb-a1b7-bc8d34971a22.png">



> * Is it possible to check if `buildx` is installed/enabled?

I don't know.



> * If not, can it be installed on the fly?

I think yes

**Reply by aalmiray** [reply]
Filed https://github.com/jreleaser/jreleaser/issues/1046

**Reply by aalmiray** [reply]
A particular issue when building multiarch images is that they must be pushed to a registry as a result otherwise the work is lost AFAICT. This would mean:
- pushing the image during `package`, not `publish` as before.
- must be performed for each configured registry as it becomes part of the tag.
- no configured registry means no multiarch image -> validation error instead of warning.

**Reply by gotson** [reply]
> * pushing the image during `package`, not `publish` as before.

You can actually build without pushing, resulting in images being stored in cache. Then when you publish it will get them from cache. You could also export them as tar manually if it runs on different machines/jobs.


> * must be performed for each configured registry as it becomes part of the tag.

You can add multiple tags to a single image, even from multiple registries, and perform a single push that will push to all registries.

> * no configured registry means no multiarch image

usually if there is no registry, by default it uses DockerHub. Not sure why it would be different for a single arch image or a multi-arch image.

**Reply by aalmiray** [reply]
> > * pushing the image during `package`, not `publish` as before.
> 
> You can actually build without pushing, resulting in images being stored in cache. Then when you publish it will get them from cache. You could also export them as tar manually if it runs on different machines/jobs.

Oh goodie, this would keep `package` and `publish` as they are right.

> > * must be performed for each configured registry as it becomes part of the tag.
> 
> You can add multiple tags to a single image, even from multiple registries, and perform a single push that will push to all registries.

Yes, except you must login for each registry separately, right? Thus multiple pushes, one per registry I suppose.

> > * no configured registry means no multiarch image
> 
> usually if there is no registry, by default it uses DockerHub. Not sure why it would be different for a single arch image or a multi-arch image.

Currently JReleaser assumes no registry means nothing to be published. This could be changed to point to DockerHub as default.

**Reply by gotson** [reply]
> Yes, except you must login for each registry separately, right? Thus multiple pushes, one per registry I suppose.
> 
> > >

Indeed you need multiple logins, but a single push is enough.

**Reply by gotson** [reply]
> You can actually build without pushing, resulting in images being stored in cache. Then when you publish it will get them from cache. You could also export them as tar manually if it runs on different machines/jobs.
> 
> >

Here is the output of `docker buildx build` if you don't add `--push`: `WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load`

**Reply by aalmiray** [reply]
Indeed. I can see the cached entries in this command

```$ docker -l error buildx build --platform linux/amd64,linux/arm64/v8  -f /Users/aalmiray/tmp/app/out/jreleaser/package/app/docker/Dockerfile --push -t aalmiray/app:v1.0.0 /Users/aalmiray/tmp/app/out/jreleaser/package/app/docker
[+] Building 6.9s (15/15) FINISHED                                                                                                                                                                                                            
 => [internal] load build definition from Dockerfile                                                                                                                                                                                     0.0s
 => => transferring dockerfile: 887B                                                                                                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                        0.0s
 => => transferring context: 2B                                                                                                                                                                                                          0.0s
 => [linux/arm64 internal] load metadata for docker.io/azul/zulu-openjdk:19-jre                                                                                                                                                          1.1s
 => [linux/amd64 internal] load metadata for docker.io/azul/zulu-openjdk:19-jre                                                                                                                                                          1.4s
 => [linux/amd64 1/4] FROM docker.io/azul/zulu-openjdk:19-jre@sha256:15a9c7a45dd9633cec8cd8db30cc57e98805fe71c8334013b4111077f1c644b0                                                                                                    0.0s
 => => resolve docker.io/azul/zulu-openjdk:19-jre@sha256:15a9c7a45dd9633cec8cd8db30cc57e98805fe71c8334013b4111077f1c644b0                                                                                                                0.0s
 => [linux/arm64 1/4] FROM docker.io/azul/zulu-openjdk:19-jre@sha256:15a9c7a45dd9633cec8cd8db30cc57e98805fe71c8334013b4111077f1c644b0                                                                                                    0.0s
 => => resolve docker.io/azul/zulu-openjdk:19-jre@sha256:15a9c7a45dd9633cec8cd8db30cc57e98805fe71c8334013b4111077f1c644b0                                                                                                                0.0s
 => [internal] load build context                                                                                                                                                                                                        0.0s
 => => transferring context: 104B                                                                                                                                                                                                        0.0s
 => CACHED [linux/arm64 2/4] RUN apt-get update && apt-get install -y unzip binutils                                                                                                                                                     0.0s
 => CACHED [linux/arm64 3/4] COPY assembly/ /                                                                                                                                                                                            0.0s
 => CACHED [linux/arm64 4/4] RUN mkdir -p /app-1.0.0/bin &&     mkdir -p /app-1.0.0/lib &&     mv /app /app-1.0.0/bin &&     chmod +x /app-1.0.0/bin/app &&     mv /app-1.0.0.jar /app-1.0.0/lib                                         0.0s
 => CACHED [linux/amd64 2/4] RUN apt-get update && apt-get install -y unzip binutils                                                                                                                                                     0.0s
 => CACHED [linux/amd64 3/4] COPY assembly/ /                                                                                                                                                                                            0.0s
 => CACHED [linux/amd64 4/4] RUN mkdir -p /app-1.0.0/bin &&     mkdir -p /app-1.0.0/lib &&     mv /app /app-1.0.0/bin &&     chmod +x /app-1.0.0/bin/app &&     mv /app-1.0.0.jar /app-1.0.0/lib                                         0.0s
 => exporting to image                                                                                                                                                                                                                   5.5s
 => => exporting layers                                                                                                                                                                                                                  0.1s
 => => exporting manifest sha256:83415aed500eba4eaacdbd1f2bdf0c6bfca8f3810608a7594222c1ee659a65a0                                                                                                                                        0.0s
 => => exporting config sha256:f18a6907001c09e7600ce26ab770a20180cd0fad71bd35ead9b8d613754c24e7                                                                                                                                          0.0s
 => => exporting manifest sha256:a4c940151497bff6c2947542a5490a21168283cb87ed0b4c0f9c0ae653d3d12c                                                                                                                                        0.0s
 => => exporting config sha256:e08ff26b9f2b866d1f252828798c923e9d8dce8f3d0a366edc408a7ec79f7ce0                                                                                                                                          0.0s
 => => exporting manifest list sha256:73e4fca8add609188cdc4a8ca4645db4adb22f5f6611b440169db11b34abf8df                                                                                                                                   0.0s
 => => pushing layers                                                                                                                                                                                                                    3.2s
 => => pushing manifest for docker.io/aalmiray/app:v1.0.0@sha256:73e4fca8add609188cdc4a8ca4645db4adb22f5f6611b440169db11b34abf8df                                                                                                        2.2s
 => [auth] aalmiray/app:pull,push token for registry-1.docker.io                                                                                                                                                                         0.0s
```

I suppose multiple tags may be specified when building, say `-t <registry1>/<username>/<repo>:<tag>,<registry2>/<username>/<repo>:<tag>` given "registry1", and "registry2". This would allow building a single image with multiple tags. Next, when it comes to pushing tags must be split between registries so that login happens once per registry, such that

```
$ docker login <registry1> -u <username> -p <password>
$ docker buildx build --platform linux/amd64,linux/arm64/v8  -f Dockerfile --push -t <registry1>/<username>/<repo>:<tag> .
$ docker login <registry2> -u <username> -p <password>
$ docker buildx build --platform linux/amd64,linux/arm64/v8  -f Dockerfile --push -t <registry2>/<username>/<repo>:<tag> .
```

Am I correct in assuming this would work?

**Reply by gotson** [reply]
You can actually prepare and push with different tags, and you can login first to multiple registries before pushing.

Example in this job:
- [login to DockerHub](https://github.com/gotson/komga/actions/runs/3375744673/jobs/5602757552#step:8:11)
- [login to ghcr.io](https://github.com/gotson/komga/actions/runs/3375744673/jobs/5602757552#step:9:12)
- the push of [layers](https://github.com/gotson/komga/actions/runs/3375744673/jobs/5602757552#step:10:1212) and [manifest](https://github.com/gotson/komga/actions/runs/3375744673/jobs/5602757552#step:10:1214) is separate with buildx (not sure about classic docker). You can see in the manifest push that multiple tags are pushed at once.

This is [my script](https://github.com/gotson/komga/blob/master/ci/publish-dockerhub.sh#L10-L20) for pushing multiple tags.

**Reply by gotson** [reply]
I'm pretty sure tagging during image build is not necessary, and only required when pushing. For instance my [prepare script](https://github.com/gotson/komga/blob/master/ci/prepare-dockerhub.sh#L13-L19) doesn't tag for ghcr.io, and [the push](https://github.com/gotson/komga/blob/master/ci/publish-dockerhub.sh#L10-L20) still works fine.

**Reply by aalmiray** [reply]
Oh. Interesting. Login to all registries beforehand and issuing a single push should make the release faster.

**Reply by aalmiray** [reply]
Success! These images were published using `buildx` support with a local snapshot.

https://hub.docker.com/repository/docker/aalmiray/app/tags?page=1&ordering=last_updated
